Scheduler does not check permissions on server_dyn_res script

arungrover opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
PBS does not check for permissions of server/mom dynamic resource script. There can be a case that the permissions of this file are open for 'write' to others and group.

#### Affected Platform(s)
All

#### Cause / Analysis / Design
No permission checks were performed.


#### Solution Description
Added permission checks and also added a check that the user owning the script should have a uid <10 or be an admin (on windows)

#### Testing logs/output
[test_momdynres.txt](https://github.com/PBSPro/pbspro/files/2631339/test_momdynres.txt)
[test_svrdynres.txt](https://github.com/PBSPro/pbspro/files/2631340/test_svrdynres.txt)
[test_smoketest.txt](https://github.com/PBSPro/pbspro/files/2631342/test_smoketest.txt)


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

bhroam commented on test/tests/functional/pbs_mom_dynamic_resource.py:
```diff
@@ -365,24 +365,24 @@ def test_mom_dyn_res_permissions(self):
+        self.check_access_log(fp, exist=False)
 
-        # Cleanup
-        os.remove(fp)
```
You want to add this to the file list as close to the create_temp_file() as possible.

bhroam commented on test/tests/functional/pbs_mom_dynamic_resource.py:
```diff
@@ -427,13 +427,13 @@ def test_mom_dyn_res_permissions(self):
-        # Cleanup
-        os.remove(fp)
+        # Add to filenames for cleanup
+        self.filenames.append(fp)
```
Same here, move it to after the create_temp_file()

bhroam commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -652,19 +652,19 @@ def test_svr_dyn_res_permissions(self):
+        self.check_access_log(fp)
 
         # Cleanup
         os.remove(fp)
```
Add this to the file list instead of removing it (up near create_temp_file())

arungrover commented on test/tests/functional/pbs_mom_dynamic_resource.py:
```diff
@@ -391,19 +391,19 @@ def test_mom_dyn_res_permissions(self):
+        self.check_access_log(fp)
 
         # Cleanup
         os.remove(fp)
```
Well, in this case file is created inside a temporary directory, I guess I'll change teardown to handle directories as well

arungrover commented:
Thanks for reviewing the code @bhroam. I've addressed your comments, please have a look

agrawalravi90 commented on src/resmom/mom_main.c:
```diff
@@ -3429,8 +3432,20 @@ add_static(char *str, char *file, int linenum)
+		err = tmp_file_sec(&str[1], 0, 1, S_IWGRP|S_IWOTH, 1);
+#endif
+		if (err != 0) {
+			snprintf(log_buffer, sizeof(log_buffer), "error: %s file has a non-secure file access, errno: %d", &str[1], err);
```
this looks like a very long line, can you please break it out into 2?

agrawalravi90 commented on src/resmom/mom_main.c:
```diff
@@ -3429,8 +3432,20 @@ add_static(char *str, char *file, int linenum)
+#endif
+		if (err != 0) {
+			snprintf(log_buffer, sizeof(log_buffer), "error: %s file has a non-secure file access, errno: %d", &str[1], err);
+			log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_SERVER, LOG_DEBUG, __func__, log_buffer);
```
There's a 'PBSEVENT_SECURITY' event type, should we be logging at that? I also feel that LOG_DEBUG might be too low, shouldn't it be LOG_ERR?

agrawalravi90 commented on src/scheduler/parse.c:
```diff
@@ -739,8 +739,28 @@ parse_config(char *fname)
 								tok++;
 								tmp2 = string_dup(tok);
+								filename = tmp2;
+								/* skip any command line args (if any) */
```
Nit: you used 'any' twice, you don't need the first one

mike0042 commented on pbspro.spec:
```diff
@@ -397,22 +397,6 @@ if [ "$1" != "1" ]; then
-if [ -r %{pbs_prefix}/libexec/pbs_init.d ]; then
-	install -D %{pbs_prefix}/libexec/pbs_init.d /etc/init.d/pbs
-fi
-
```
If you change pbspro.spec, you must update pbspro.spec.in (or the other way around).

mike0042 commented on src/scheduler/server_info.c:
```diff
@@ -700,13 +700,37 @@ query_server_dyn_res(server_info *sinfo)
 #endif
 
-	for (i = 0; i < MAX_SERVER_DYN_RES && conf.dynamic_res[i].res != NULL; i ++) {
+	for (i = 0; i < MAX_SERVER_DYN_RES && conf.dynamic_res[i].res != NULL; i++) {
```
Please add parenthesis around each conditional.

arungrover commented:
@mike0042 and @agrawalravi90  thanks for catching the pbspro.spec change. I didn't notice it when I raised the pull. 
It seems that change was not something I did intentionally. That change had crept in because pbspro.spec and pbspro.spec.in file do not match. There was a change added in pull #876 which never made it to pbspro.spec.in file. This is why whenever I run configure in my development environment, it generates this new pbspro.spec file which does not have these lines of code.

I tried changing pbspro.spec.in file so that it has the changes #876 introduced but after that change pbs_habitat is failing. So for now, I have changed pbspro.spec to reflect what 'master' has and I'll let Prakash know about this mismatch.

agrawalravi90 commented on test/tests/functional/pbs_mom_dynamic_resource.py:
```diff
@@ -338,8 +349,124 @@ def test_alter_res_value(self):
+        self.du.chmod(path=fp, mode=0744)
+        self.check_access_log(fp, exist=False)
+
+        # Create a script in home directory which has more open privileges
```
You are creating the script in a directory that's inside the home directory. I think the comment should just mention that you are creating the script in a directory that is not secure

agrawalravi90 commented on test/tests/functional/pbs_mom_dynamic_resource.py:
```diff
@@ -338,8 +349,124 @@ def test_alter_res_value(self):
+
+        # Create dynamic resource script in tmp directory and check
+        # file permissions
+        fp = self.du.create_temp_file(body=scr_body, dirname=home_dir)
```
You are creating the script in home directory and not tmp, right? How is this section of checks different from the checks that you've already performed earlier in this test?

arungrover replied:
Ahh... that dirname should not be there. I'll remove it

agrawalravi90 commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -48,6 +48,23 @@ def setUp(self):
+        Helper function to check if scheduler logged a file security
+        message.
+        """
+        match_from = time.time()
```
you haven't int()'d this yet

agrawalravi90 commented on test/tests/functional/pbs_mom_dynamic_resource.py:
```diff
@@ -70,6 +70,17 @@ def create_mom_resources(self, resc_name, resc_type, resc_flag,
         return fp_list
 
+    def check_access_log(self, fp, exist=True):
+        match_from = time.time()
```
if you decide to int() the value in server_dyn_res test, then please int() this value as well

agrawalravi90 commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -588,3 +607,122 @@ def test_res_boolean_invalid_input(self):
+                                 starttime=match_from, existence=True,
+                                 max_attempts=10)
+
+        # Create dynamic resource script in tmp directory and check
```
Same comment here, can you please explain what additional testing this section of the test will perform?

arungrover commented on test/tests/functional/pbs_mom_dynamic_resource.py:
```diff
@@ -338,8 +349,124 @@ def test_alter_res_value(self):
+
+        # give write permission to user only
+        self.du.chmod(path=fp, mode=0744)
+        self.check_access_log(fp, exist=False)
```
Well, it is passed to log_match() function in an arg names 'existence' which is on the same lines. If you are ok, I'd like to retain it.

arungrover commented:
@agrawalravi90 @bhroam I've addressed more review comments. In the PTL tests check_access_log() was wrongly capturing time before sleeping, I moved that after sleep as well.

agrawalravi90 commented:
Thanks for the changes Arun. It looks find to me as such, I'm just still not clear on why you are testing whether creating the script in tmp works or not. You've already tested creating the script in a directory with open access, isn't that similar to creating it in tmp? I feel like you only need one of them, either test for tmp, or for a custom directory with open access, what do you think?

arungrover commented:
Hey Ravi, there is a subtle difference between the tests. Temp directory has sticky bit set on it and has open permissions. This part of the test tests that tmp directory with open permissions is permissible as long as file has correct permission.

agrawalravi90 commented:
Thanks for clarifying Arun, I sign-off on the changes as well, please rebase in preparation for merge

arungrover commented:
Thanks Ravi
I'll rebase it soon. I have to test the change on windows once.

agrawalravi90 commented:
@arungrover Any update on this?

arungrover commented:
@agrawalravi90 I found a problem on windows, still working on it

arungrover commented:
I have added some more changes because the code was broken on windows.
I have tested it on windows now and it runs fine in the positive scenario where permissions are not open, But unfortunately, tmp_file_sec() function seems to be broken on windows and always returns SUCCESS. 
I am not sure how file permissions behave on windows and thus I will ask windows gurus (hint: @bremanandjk and @hirenvadalia ) to have a look at it. For now, please review the rest of it and provide comments.

hirenvadalia commented:
@arungrover However I don't know the full story here but since you mentioned `tmp_file_sec()` so I suggest you look whether `teststat()` and `perm_granted_admin_and_owner()` are doing a correct job or not.

arungrover commented:
Well I did glanced through the code and it seems to be doing the right thing but I’m not an expert on windows and that is why I tagged.
Problem is that this function is supposed to return an error when a non admin own the file but they don’t.  Like chk_tree binary on windows, it does the same thing but I could not get that binary to fail stating permission issues.

agrawalravi90 commented:
@arungrover any luck with the Windows issues?

arungrover commented:
Well it runs on windows but tmp file sec function always seems to return false positive results. I'll talk to hiren and brem tomorrow again and see if they know something about it. In the meantime I would suggest that we continue with the review. Even if the function to check for file security turns out to be faulty on windows we can address that separately since it is already being used in other places.

agrawalravi90 commented on src/lib/Liblog/chk_file_sec.c:
```diff
@@ -452,3 +452,77 @@ tmp_file_sec(char *path, int isdir, int sticky, int disallow, int fullpath)
+		}
+		else if (S_ISREG(sbuf.st_mode)) {
+			path_exists = 1;
+			prev_space = next_space;
```
I must be missing something, why do you not break out of the loop at this point? you've found the program name right?

agrawalravi90 replied:
Got it. Can you please add a comment in the code to explain this?

agrawalravi90 commented on src/lib/Liblog/chk_file_sec.c:
```diff
@@ -452,3 +452,77 @@ tmp_file_sec(char *path, int isdir, int sticky, int disallow, int fullpath)
+		return tok;
+	}
+
+	/* If control is here then it would mean that "tok" must have file path*/
```
tok could have had a file path even before control reached here right?

agrawalravi90 replied:
I see, so maybe the comment should be "/* If control is here then it would mean that "tok" must have **only** file path*/" ?

agrawalravi90 commented on src/scheduler/data_types.h:
```diff
@@ -970,6 +970,7 @@ struct dyn_res
 {
 	char *res;
 	char *program;
+	char *script;
```
how is 'script' different from 'program' ?

arungrover replied:
Program consists of arguments which is given directly to popen command. Script just conatins the name of the script. I did this because I did not want scheduler to search for script name every time it executes it.

agrawalravi90 replied:
i'd also stick to either 'program' or 'script', using both makes it sound like program is different from script.
How about 'command_line' and 'script_name' ?

bhroam replied:
+1 @agrawalravi90's names

agrawalravi90 commented on test/tests/functional/pbs_mom_dynamic_resource.py:
```diff
@@ -388,13 +388,13 @@ def test_mom_dyn_res_permissions(self):
         # Create script in a directory which has more open privileges
         # This should make loading of this file fail in all cases
-        dir_temp = self.du.mkdtemp(mode=0766, dir=home_dir)
+        dir_temp = self.du.mkdtemp(mode=0766, dir=home_dir, suffix=' tmp')
```
there is a space in the suffix string, is that intentional?

arungrover replied:
It is, my program failed when I ran it on windows because I had my script residing in "Program Files (x86)" which had spaces and the logic before this change interpreted script name as "Program". That is why I changed the test to add a space in the directory name.

agrawalravi90 replied:
so you added the space because you wanted to test that your changes work even when the script is located in a folder which has spaces in its name? If that's the case then can you please add a comment here saying that you are adding the space for this reason?

arungrover replied:
Sure, I'll add a comment

mike0042 commented on pbspro.spec:
```diff
@@ -397,22 +397,6 @@ if [ "$1" != "1" ]; then
 	echo
 fi
 
-%posttrans %{pbs_server}
```
I agree with @agrawalravi90. Please rebase your branch. This change should not be required.

arungrover replied:
ahh... I'm sorry this change again crept in. I usually push after doing a build and since this is an autogenerated file it becomes part of the commit. I'll rebase in my next push.

arungrover commented on src/lib/Liblog/chk_file_sec.c:
```diff
@@ -452,3 +452,77 @@ tmp_file_sec(char *path, int isdir, int sticky, int disallow, int fullpath)
+		tok[strlen(tok)-1] = '\0';
+
+	next_space = strpbrk(tok, delim);
+	for (; next_space != NULL; next_space = strpbrk(next_space, delim)) {
```
When this function is called, the input string should start with the program name. in this loop we are finding if there are any space separated arguments to the program (or the name itself has space in it)

arungrover commented on src/lib/Liblog/chk_file_sec.c:
```diff
@@ -452,3 +452,77 @@ tmp_file_sec(char *path, int isdir, int sticky, int disallow, int fullpath)
+			prev_space = next_space;
+		}
+		*next_space = ' ';
+		next_space += strspn(next_space, delim);
```
I'll add a comment.

arungrover commented:
@agrawalravi90 I have addressed your review comments in this pull.
@mike0042 I did a rebase so that I do not get spec file changes again. While doing so I found that pbspro.spec and pbspro.spec.in are again out of sync. spec file now has an extra line "%exclude %{pbs_prefix}/libexec/pbs_schema_upgrade" which is missing from pbspro.spec.in

mike0042 commented:
Would you mind fixing the spec file issue @arungrover?

arungrover commented:
@mike0042 I added another commit to address it

mike0042 commented:
Thanks @arungrover!

bhroam commented on src/include/log.h:
```diff
@@ -121,6 +121,7 @@ extern int  tmp_file_sec(char *path, int isdir, int sticky,
 extern int  chk_file_sec2(char *path, int isdir, int sticky,
 	int disallow, int fullpath, char *owner);
 #endif
+extern char* get_script_name(char *input);
```
Please attach the * to the function name

bhroam commented on src/lib/Liblog/chk_file_sec.c:
```diff
@@ -452,3 +452,77 @@ tmp_file_sec(char *path, int isdir, int sticky, int disallow, int fullpath)
+ * @retval - NULL when no valid program name is found
+ * @retval - a newly allocated program name
+ *
+ * @NOTE - Caller holds the responsibility of freeing up the return value
```
I don't think the doxygen tag is capitalized

bhroam commented on src/lib/Liblog/chk_file_sec.c:
```diff
@@ -452,3 +452,77 @@ tmp_file_sec(char *path, int isdir, int sticky, int disallow, int fullpath)
+		return NULL;
+
+	/* get rid of double quotes at the end */
+	if (starts_with_quotes && tok[strlen(tok)-1] == '\"')
```
Please put spaces around the '-' sign.

bhroam commented on src/lib/Liblog/chk_file_sec.c:
```diff
@@ -452,3 +452,77 @@ tmp_file_sec(char *path, int isdir, int sticky, int disallow, int fullpath)
+		tok[strlen(tok)-1] = '\0';
+
+	next_space = strpbrk(tok, delim);
+	for (; next_space != NULL; next_space = strpbrk(next_space, delim)) {
```
You are initializing next_space outside of the for() loop.  Why not do it in the initialization portion?

arungrover commented:
@agrawalravi90 and @bhroam  thanks for reviewing the pull, I've posted another commit based on the review comments.

agrawalravi90 commented:
Also, please post latest test logs after you rebase

arungrover commented:
[test_svrdynres.txt](https://github.com/PBSPro/pbspro/files/2761490/test_svrdynres.txt)
[test_momdynres.txt](https://github.com/PBSPro/pbspro/files/2761493/test_momdynres.txt)

anamikau commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -11007,14 +11007,16 @@ def add_server_dyn_res(self, custom_resource, script_body=None,
                          configuration settings.
         :type validate: bool
         """
+        super_user_grp = PbsGroup(grp.getgrgid(0).gr_name, gid=0) 
```
should we be making this change in ptl/utils/pbs_testsuite.py generically?

arungrover replied:
Actually, pbs_testsuite does have this lined up nicely as "ROOT_USER" but unfortunately, pbs_testlib has no access to it. I didn't want to complicate it by making it visible to pbs_testlib so I just replicated the code

arungrover replied:
Bhroam and I had some more discussion about it and we thought that passing a string would be better. I've changed it to a string now.

agrawalravi90 commented:
Still looks good to me, please rebase in prep for merge

arungrover commented:
@agrawalravi90 I'm facing issues running these tests as non-root users because they create dynamic resource scripts that are not root owned. This change is turning out to be bigger than I expected.

agrawalravi90 commented:
@arungrover just add a skip if the user is not root? Look at pbs_root_owned_script.py

arungrover commented:
@agrawalravi90 and @bhroam  I've made changes to the tests and PTL framework to make sure that the dynamic resource script is created as root irrespective of which user is running it. Please have a look

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -11007,22 +11007,24 @@ def add_server_dyn_res(self, custom_resource, script_body=None,
         self.logger.info("-" * 30)
-
-        self.du.chmod(self.hostname, path=res_file, mode=0755)
+        if file_specified is True:
```
you can just say "if file_specified"

agrawalravi90 replied:
also, why not just do the chmod as part of the "if res_file is not None" block?

arungrover replied:
You are right, it was silly of me to not do that instead

agrawalravi90 replied:
Ah, ok, i guess it's just personal preference then

bhroam replied:
I take it back.  I checked the pep-8 style guide and they say:
Yes:   if greeting:
No:    if greeting == True:
Worse: if greeting is True:

I guess I'll be changing how I code with booleans now.

agrawalravi90 commented on test/fw/ptl/utils/pbs_dshutils.py:
```diff
@@ -1968,6 +1968,25 @@ def mkdtemp(self, hostname=None, suffix='', prefix='PtlPbs', dir=None,
+        """
+        Create a dynamic resource script owned by root
+        """
+        home_dir = os.path.expanduser("~")
```
if this file is supposed to be owned by root, then shouldn't you create it in root's home directory instead of the current user's ?

arungrover replied:
Since while testing pbs_benchpress runs as pbsroot I don't think this user will have privileges to create a file in root's home directory. This is why I create it in user's home and copy it as root to temp directory (or any other specified directory)

agrawalravi90 commented on test/fw/ptl/utils/pbs_dshutils.py:
```diff
@@ -1968,6 +1968,25 @@ def mkdtemp(self, hostname=None, suffix='', prefix='PtlPbs', dir=None,
+        fp = self.create_temp_file(prefix="mom_resc", suffix=".scr",
+                                   body=body, asuser="root",
+                                   dirname=home_dir, hostname=host)
+        self.chmod(path=fp, mode=perm, sudo=True, hostname=host)
```
create_temp_file should probably just take a 'mode' argument, it does a chmod internally with a hardcoded 644, so it would be easy to just have it take the 'mode' argument, what do you think?

agrawalravi90 commented on test/fw/ptl/utils/pbs_dshutils.py:
```diff
@@ -1968,6 +1968,25 @@ def mkdtemp(self, hostname=None, suffix='', prefix='PtlPbs', dir=None,
+                                   body=body, asuser="root",
+                                   dirname=home_dir, hostname=host)
+        self.chmod(path=fp, mode=perm, sudo=True, hostname=host)
+        if dirname is None:
```
why not just directly pass dirname to create_temp_file ?

agrawalravi90 commented on test/tests/functional/pbs_mom_dynamic_resource.py:
```diff
@@ -373,28 +377,33 @@ def test_mom_dyn_res_permissions(self):
-        self.du.chmod(path=fp, mode=0744)
-        self.check_access_log(fp, exist=False)
+        self.du.chmod(path=fp, mode=0744, sudo=True)
+        root_priv = self.du.run_cmd(
```
if you just want to check that the directory exists then why not do os.path.isdir()?

arungrover replied:
I'm trying to check if the user running the test is root or not. I'll change it to do os.getuid() instead

agrawalravi90 commented on test/tests/functional/pbs_mom_dynamic_resource.py:
```diff
@@ -403,80 +412,59 @@ def test_mom_dyn_res_permissions(self):
 
         # give write permission to others
-        self.du.chmod(path=fp, mode=0746)
+        fp = self.du.create_dyn_res_script(scr_body, perm=0746)
```
why are you creating a new file every time?

agrawalravi90 commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -214,9 +213,9 @@ def test_res_whitespace(self):
-        self.du.chmod(path=fn_in, mode=0755, sudo=True)
+        fn_in = self.du.create_dyn_res_script(script_body,
+                                              prefix="PtlPbs_get_foo")
+        self.filenames.append(fn_in)
```
just curious, did previously the temp files just stick around after running the test?

arungrover replied:
I was just making sure that teardown of each test clear files created by them. This is because PTL's cleanup is dependent on the file name prefixed with PtlPbs* and there are bunch of tests that creates files with different prefixes. So I'm going to do cleanup for those files, might as well do it for others.

agrawalravi90 commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -256,22 +257,18 @@ def test_multiple_res(self):
+        fn_l = self.du.create_dyn_res_script(script_body_l,
+                                             prefix="PtlPbs_large",
+                                             suffix=".scr")
+        self.filenames.append(fn_s)
```
this might be a bit of work, but it would have been much much cleaner if there was a 'globals' module in PTL, which had a list, say, "file_to_cleanup" defined, which functions like create_dyn_res_script could add filenames to so that at the end of the test tearDown() could check the list and delete the files. This would remove the need for a test writer to remember to clean up the files that they create.

bhroam replied:
There's an existing bug out there for this.  I filed it a few months ago.  If you name your files correctly, PbsTestSuite will clean up for you.  Look at test_runner().  It will clean up after the test

agrawalravi90 commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -685,52 +681,44 @@ def test_svr_dyn_res_permissions(self):
-                os.remove(i)
-            elif os.path.isdir(i):
-                os.removedirs(i)
+        self.du.rm(path=self.filenames, sudo=True, force=True,
```
you should probably also call PTL's tearDown() if you are going to replace it

arungrover replied:
oops! I'll add it

bremanandjk commented:
> I have added some more changes because the code was broken on windows.
> I have tested it on windows now and it runs fine in the positive scenario where permissions are not open, But unfortunately, tmp_file_sec() function seems to be broken on windows and always returns SUCCESS.
> I am not sure how file permissions behave on windows and thus I will ask windows gurus (hint: @bremanandjk and @hirenvadalia ) to have a look at it. For now, please review the rest of it and provide comments.

Thanks @arungrover for looping me 
I have raised a PR to fix this file security issue, plz share your comments.
PR link: https://github.com/PBSPro/pbspro/pull/990

arungrover commented:
Hi @agrawalravi90 and @bhroam I've pushed my latest changes. most of the changes are in PTL Fw and tests. I'd also request @kjakkali to have a look at the modified create_temp_file() method.

I had to rebase and force push my changes because it was not building off of the old master.

agrawalravi90 commented:
@arungrover can you please mention which of your new commits are really "new" so that i can just review those? thanks

arungrover commented:
Ah sorry, I should have mentioned that earlier. - It's https://github.com/PBSPro/pbspro/pull/905/commits/25b4181fdc60dd634a9c045ac21fd8df055a0bf1

bhroam commented:
Travis failed due to pep8.  Codacy is mostly complaining about stuff that we don't care about.  There is one variable that is set and never used.

kjakkali commented on test/fw/ptl/utils/pbs_dshutils.py:
```diff
@@ -1876,10 +1876,12 @@ def create_temp_file(self, hostname=None, suffix='', prefix='PtlPbs',
         :type body: str or None
         :param level: logging level, defaults to INFOCLI2
         :type level: int
+        :param mode: mode to use while creating directories
```
Please update the doc . we are not creating dir here.

arungrover commented on test/tests/functional/pbs_mom_dynamic_resource.py:
```diff
@@ -361,35 +357,35 @@ def test_mom_dyn_res_permissions(self):
 
         # give write permission to group and others
-        self.du.chmod(path=fp, mode=0766, sudo=True)
+        self.du.chmod(path=fp, mode=0766, sudo=True, runas=ROOT_USER)
```
You are right, I was trying multiple things out within dshutils and left this cruft here.

arungrover commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -13451,6 +13471,57 @@ def action(self, act, script):
+            script_body = f.readlines()
+            f.close()
+            self.du.chmod(self.hostname, path=res_file, mode=perm, sudo=True)
+        else:
```
I'm not sure if I should do that, there might be a case where a test might want to test the functionality with an empty script.  I can probably not give a default value to script_body if you like it that way.

arungrover commented on test/fw/ptl/utils/pbs_dshutils.py:
```diff
@@ -1892,29 +1892,36 @@ def create_temp_file(self, hostname=None, suffix='', prefix='PtlPbs',
-                          preserve_permission=False, level=level)
-            os.unlink(tmpfile)
+                          preserve_permission=False, level=level, mode=mode)
+            if dest_file != tmpfile:
```
The outer condition is a set of 3 OR conditions, it might happen that we enter into this block because of some other condition.

arungrover commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -89,12 +90,13 @@ def setup_dyn_res(self, resname, restype, resval):
             # server_dyn_res entry.
             # HACK: So adding a single resource first and then the list.
```
I think it isn't. Thanks for finding it out, I'll remove it.

arungrover commented on test/tests/functional/pbs_svr_dyn_res.py:
```diff
@@ -89,12 +90,13 @@ def setup_dyn_res(self, resname, restype, resval):
+                                                          suffix=".scr")
+            val.append('"' + resname[i] + ' ' + '!' + dest_file + '"')
+            self.filenames.append(dest_file)
+        a = {'server_dyn_res': val}
```
I think it just created the value of server_dyn_res in the loop.

arungrover commented:
@agrawalravi90, @bhroam @kjakkali I've addressed your comments. Please have a look.
Please Ignore my force push as I realized I missed one change and then I did that change and squashed it into the last commit. commit id 
58e4332 has the change.

test logs - 
[test_ptl.txt](https://github.com/PBSPro/pbspro/files/2913579/test_ptl.txt)
[test_mom.txt](https://github.com/PBSPro/pbspro/files/2913580/test_mom.txt)
[test_svr.txt](https://github.com/PBSPro/pbspro/files/2913581/test_svr.txt)
